pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+    and then to a :class:`~pytential.symbolic.primitives.IntG`.
+    """
+    from pytential.symbolic.pde.system_utils import simplify_densities
+    to_pymbolic = SympyToPymbolicMapper()
+
+    orig_kernel = orig_int_g.source_kernels[0]
+    source_kernels = []
+    densities = []
+    for monom, coeff in poly.terms():
+        kernel = orig_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisSourceDerivative(idim, kernel)
+        source_kernels.append(kernel)
+        densities.append(to_pymbolic(coeff) * (-1)**sum(monom))
+    return orig_int_g.copy(source_kernels=tuple(source_kernels),
+            densities=tuple(simplify_densities(densities)))
+
+
+def _convert_target_poly_to_int_g(poly, orig_int_g, rhs_int_g):
+    """This does the opposite of :func:`_kernel_source_derivs_as_poly`
+    and converts a polynomial back to a target derivative
+    operator. It is applied to a :class:`~pytential.symbolic.primitives.IntG`
+    object and returns a new instance.
+    """
+    to_pymbolic = SympyToPymbolicMapper()
+
+    result = 0
+    for monom, coeff in poly.terms():
+        kernel = orig_int_g.target_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisTargetDerivative(idim, kernel)
+        result += orig_int_g.copy(target_kernel=kernel,
+                source_kernels=rhs_int_g.source_kernels,
+                densities=rhs_int_g.densities) * to_pymbolic(coeff)
+
+    return result
+
+
+class CoefficientCollector(Mapper):
+    """From a density expression, extracts expressions that need to be
+    evaluated for each source and coefficients for each expression.
+
+    For eg: when this mapper is given as ``s*(s + 2) + 3`` input,
+    it returns {s**2: 1, s: 2, 1: 3}.
+    """
+    def __init__(self, source_dependent_variables):
+        self.source_dependent_variables = source_dependent_variables
+
+    def __call__(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        return super().__call__(expr)
+
+    def map_sum(self, expr):
+        stride_dicts = [self.rec(ch) for ch in expr.children]
+
+        result = defaultdict(lambda: 0)
+        for stride_dict in stride_dicts:
+            for var, stride in stride_dict.items():
+                result[var] += stride
+        return dict(result)
+
+    def map_algebraic_leaf(self, expr):
+        if expr in self.source_dependent_variables:
+            return {expr: 1}
+        else:
+            return {1: expr}
+
+    def map_subscript(self, expr):
+        if expr in self.source_dependent_variables or \
+                expr.aggregate in self.source_dependent_variables:
+            return {expr: 1}
+        else:
+            return {1: expr}
+
+    def map_constant(self, expr):
+        return {1: expr}
+
+    def map_product(self, expr):
+        if len(expr.children) > 2:
```

Comment:
Add section comment: adjust nesting to ensure exactly two children.

Reply:
Done in 0994482d89465d36566f2d5345d68f528eb0b739